Prometheus: Metrics browser

davkal opened the pull request:
Similarly to the Logs browser for Loki (#30351), this PR adds a Metrics browser for Prometheus.

![Kapture 2021-05-10 at 17 35 32](https://user-images.githubusercontent.com/859729/117685423-3669ce00-b1b6-11eb-90a7-65061b416e08.gif)



Notes for reviewer:

Ideally this still makes it into the beta 8. The metric (`__name__`) are a special label and are treated specially. As a result, I eschewed modifying the LogLabelBrowser, since this would have meant conditional code all over the place. I'd like to revisit this decision later, and either make it composable, or extract more common logic. 

What's missing now (compared to the Metrics cascader) is the Histogram and Recording rules groups, see Todos below.

Todos for later PRs (**moved to epic: #33976**):

- [x] add HELP and TYPE
- [ ] make metrics collapsible
- [ ] add histograms helper (search {le=~".+"} first?)
- [ ] add recording rules helper (or filter)
- [ ] investigate the new Labels APIs that recently started taking a selector (to replace the use of the Series API)

davkal committed:
[WIP] Metrics browser

davkal committed:
Removed unused import

davkal committed:
Metrics selection logic

ifrost commented:
Previously recording rules and histograms were grouped separately:

<img width="911" alt="Screenshot 2021-05-11 at 14 55 55" src="https://user-images.githubusercontent.com/745532/117818746-31198b80-b269-11eb-8331-4cf3d4edf8b4.png">

Is there a way to access it now? There's a way to run "rate query" with a button at the bottom, but I can't find a way to get histogram query as before. Is it somewhere in the UI? Or maybe there should be an extra "run histogram query" button? (or should it be figured out automagically based on metric types)

ifrost committed:
Merge branch 'main' into davkal/metrics-browser

ifrost commented on public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx:
```diff
@@ -88,13 +90,12 @@ describe('PromQueryField', () => {
   });
 
-  it('does not refreshes metrics when after rounding to minute time range does not change', async () => {
+  it.skip('does not refreshes metrics when after rounding to minute time range does not change', async () => {
```
Why is it skipped?

ifrost commented on public/app/plugins/datasource/prometheus/components/PromQueryField.tsx:
```diff
@@ -42,13 +40,14 @@ function getChooserText(metricsLookupDisabled: boolean, hasSyntax: boolean, metr
+  return 'Metrics browser';
 }
 
+/* TODO
```
Is it still needed? Can the commented code-block be removed?

ifrost commented on public/app/plugins/datasource/prometheus/components/PromQueryField.tsx:
```diff
@@ -294,13 +284,14 @@ class PromQueryField extends React.PureComponent<PromQueryFieldProps, PromQueryF
     }
 
-    // Build metrics tree
+    /* TODO
```
Can it be removed?

ifrost commented:
> (...) As a result, I eschewed modifying the LogLabelBrowser, since this would have meant conditional code all over the place. I'd like to revisit this decision later, and either make it composable, or extract more common logic. (...)

Label.tsx and LokiLabel.tsx are almost identical. I think the only difference is the autoescape property? Could autoescape be set for Prometheus too? If yes we could make it a shared component. For the browser component - there's more tweaks in it but there's a lot of common code so it'd be great to make it a more generic component at some point. Maybe other data sources could use it too? I can imagine something like this for Graphite would be useful (and some users asked about it).

ifrost committed:
Remove redundant tests

All data is fetched now regardless to the current range so test for checking reloading the data on the range change are no longer relevant.

ifrost committed:
Remove commented out code blocks

ifrost committed:
Merge branch 'main' into davkal/metrics-browser

ifrost committed:
Add issue number to todos

ifrost reviewed (approved):
It's good to go! Added an epic to address remaining issues #33976.

oddlittlebird commented:
Is there documentation for this yet?
